### ISSUE TITLE
Load weights of transformer into PerformerLM

### ISSUE THREAD

**Issue by Mazgis47** [lucidrains/performer-pytorch#16]
Hi, is there an explanation (or even implementation) on how to load pre-trained Bert (or RoBerta or GPT-2 etc.) model weights into PerformerLM model? According to the paper "Backwards compatibility with pretrained models is available as a benefit from softmax approximation, via small finetuning (required due to error propagation)"
If I correctly understand in your readme there is a hint that SelfAttention should map to the transformer self-attention layers, but how and what finetuning is required?
e.g.
when Self attention has:
(to_qkv): Linear(in_features=768, out_features=2304, bias=False)
to_out): Linear(in_features=768, out_features=768, bias=True)
and Transformer has:
(attention): BertAttention(
        (self): BertSelfAttention(
          (query): Linear(in_features=768, out_features=768, bias=True)
          (key): Linear(in_features=768, out_features=768, bias=True)
          (value): Linear(in_features=768, out_features=768, bias=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
        (output): BertSelfOutput(
          (dense): Linear(in_features=768, out_features=768, bias=True)
          (LayerNorm): FusedLayerNorm(torch.Size([768]), eps=1e-12, elementwise_affine=True)
          (dropout): Dropout(p=0.1, inplace=False)
        )
      )
Thanks.

**Comment by lucidrains** [comment]
@Mazgis47 Hi and thanks for your interest! Indeed they appear to be different, but it's because I merge projections matrices of Q K V into one projection

i have separated them in 0.7.0, and perhaps it will be clearer to you what needs to be done

the other thing you can do is simply reach into the BertSelfAttention module and use the `FastAttention` module supplied in the repo, as detailed at the end of the Usage section https://github.com/lucidrains/performer-pytorch#usage

**Comment by lucidrains** [comment]
@Mazgis47 the only big gotcha is that with the Performer like architecture, the masking will be different. Instead of deriving the N x N mask the setting the dot product of (qk) to -infinity, you will actually simply set the tokens of the keys and values to 0 instead

**Comment by Mazgis47** [comment]
@lucidrains, thanks for reply, indeed with v. 7x it is clearler. I tried the following:
```
for i in range(len(model.performer.net.blocks)):
    model.performer.net.blocks[i].f.net.fn.to_q.weight = bert_model.bert.encoder.layer[i].attention.self.query.weight
    model.performer.net.blocks[i].f.net.fn.to_k.weight = bert_model.bert.encoder.layer[i].attention.self.key.weight
    model.performer.net.blocks[i].f.net.fn.to_v.weight = bert_model.bert.encoder.layer[i].attention.self.value.weight
    model.performer.net.blocks[i].f.net.fn.to_out.weight = bert_model.bert.encoder.layer[i].attention.output.dense.weight
``` 
Also tried with ` torch.nn.Parameter(...)` but prediction are random - every time I start over (recreate PerformerLM model and run this loop) I get different prediction results which actually never as expected like in original model. It seems that weights are taken from the moment of the creation of the PerformerLM model and not the copied ones (maybe I am wrong here).

**Comment by lucidrains** [comment]
@Mazgis47 there's a bunch of things you need to make sure are identical. i noticed that the bert selfattention linear projections have biases as well. then, you need to transfer the layer norm parameters. finally, the original bert may be using post rather than pre layer normalization. also, the predictions are random each time because performer draws a new projection matrix on each forward. you would need to do your fine-tuning and then fix the projection matrix as detailed https://github.com/lucidrains/performer-pytorch#advanced

**Comment by lucidrains** [comment]
@Mazgis47 it's random because performer is drawing a new projection matrix on each forward, but you can fix this to be deterministic after fine-tuning https://github.com/lucidrains/performer-pytorch#advanced

you should also make sure to copy over the bias values, which i noticed exists for the qkv projections on the BertSelfAttention module

**Comment by tenexcoder** [comment]
@Mazgis47 i have a working experiment that i talk about on issue #24 

Take a look and hope it helps!

**Comment by gaganbahga** [comment]
> @Mazgis47 the only big gotcha is that with the Performer like architecture, the masking will be different. Instead of deriving the N x N mask the setting the dot product of (qk) to -infinity, you will actually simply set the tokens of the keys and values to 0 instead

@lucidrains thanks for the explanation. If I understand correctly, in the code the masking is only applied by setting the values to 0, not the keys. So this would mean that the attention distribution would still be produced for the entire sequence, but the values for the masked positions would be zero. Do you think it would be helpful to set the key feature map entries for the masked positions to 0 as well? In that case the softmax kernel which is approximated by `q' k'^T`, where `q'`, `k'` are query, key random feature maps, would be zero for the masked key positions.
Or did I miss something?
